book-content/chapters/04-essential-types-and-annotations.md

```diff
@@ -0,0 +1,2366 @@
+
+You can express much more complex types in TypeScript: arrays, objects, functions and much more. We'll cover these in later chapters.
+
+### Type Inference
+
+TypeScript gives you the ability to annotate almost any value, variable or function in your code. You might be thinking “wait, do I need to annotate everything? That's a lot of extra code.”
+
+As it turns out, TypeScript can infer a lot from the context that your code is run.
+
+#### Variables Don't Always Need Annotations
+
+Let's look again at our variable annotation example, but drop the annotations:
+
+```ts twoslash
+let albumTitle = "Midnights";
+let isReleased = true;
+let trackCount = 13;
+```
+
+We didn't add the annotations, but TypeScript isn't complaining. What's going on?
+
+Try hovering your cursor over each variable.
+
+```ts twoslash
+// hovering over each variable name
+
+let albumTitle: string;
+let isReleased: boolean;
+let trackCount: number;
+```
+
+Even though they aren't annotated, TypeScript is still picking up the type that they're each supposed to be. This is TypeScript inferring the type of the variable from usage.
+
+It behaves as if we'd annotated it, warning us if we try to assign it a different type from what it was assigned originally:
+
+```ts twoslash
+// @errors: 2322
+
+let isReleased = true;
+
+isReleased = "yes";
+```
+
+And also giving us autocomplete on the variable:
+
+```ts
+albumTitle.toUpper; // shows `toUpperCase` in autocomplete
+```
+
+This is an extremely powerful part of TypeScript. It means that you can mostly _not_ annotate variables and still have your IDE know what type things are.
+
+#### Function Parameters Always Need Annotations
+
+But type inference can't work everywhere. Let's see what happens if we remove the type annotations from the `logAlbumInfo` function's parameters:
+
+```ts twoslash
+// @errors: 7006
+const logAlbumInfo = (
+  title,
+
+  trackCount,
+
+  isReleased,
+) => {
+  // rest of function body
+};
+```
+
+On its own, TypeScript isn't able to infer the types of the parameters, so it shows an error under each parameter name.
+
+This is because functions are very different to variables. TypeScript can see what value is assigned to which variable, so it can make a good guess about the type.
+
+But TypeScript can't tell from a function parameter alone what type it's supposed to be. When you don't annotate it, it defaults the type to `any` - a scary, unsafe type.
+
+It also can't detect it from usage. If we had an 'add' function that took two parameters, TypeScript wouldn't be able to tell that they were supposed to be numbers:
+
+```ts twoslash
+// @errors: 7006
+function add(a, b) {
+  return a + b;
+}
+```
+
+`a` and `b` could be strings, booleans, or anything else. TypeScript can't know from the function body what type they're supposed to be.
+
+So, when you're declaring a named function, their parameters always need annotations in TypeScript.
```

Comment:
It's probably not worth complicating it here, at this stage... but the sentence doesn't hold up to be true at all times. I could use a named function in the example I gave [here](https://github.com/Andarist/total-typescript-book/pull/4/files#r1667769572) and I could still skip annotating parameters 😉